sql: fix bugs with sequence and setval

rafiss opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/71135

See individual commits:
- sql: invalidate sequence cache when calling setval -- safe to backport
- sql: fix 2-param version of setval to use isCalled -- matches the PG behavior
- sql: make setval/nextval cross over transaction boundaries -- matches PG, and
  also fixes a possible deadlock

rafiss committed:
sql: make setval/nextval cross over transaction boundaries

Release note (bug fix): Previously, the effects of the setval and
nextval builtin functions would be rolled back if the surrounding
transaction was rolled back. This was not correct, as setval is
not supposed to respect transaction boundaries. This is fixed now.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71643)
<!-- Reviewable:end -->

rafiss committed:
sql: invalidate sequence cache when calling setval

Release note (bug fix): The setval builtin previously did not invalidate
cached sequence values. This is fixed now.

rafiss committed:
sql: fix 2-param version of setval to not set isCalled

Release note (bug fix, backward-incompatible change): The 2-parameter
setval builtin previously caused the sequence to increment incorrectly
one extra time. This is fixed now. If you do need the sequence to
increment, then use `setval(seq, val, true)`.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r1, 8 of 8 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71643#-:-MmUKlNp76kqEdyyRalI:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->
